### ISSUE TITLE
azure-iot-device v2.1.1 module not connected ('Transport timeout on connection operation')

### ISSUE THREAD

**Issue by LouanDuToitS3** [Azure/azure-iot-sdk-python#531]
Hi

I was initially very pleased with the release of v2.1.1 of the library, and it seemed that all the module to hub connectivity issues were a thing of the past. But then, unfortunately, after my pipeline has been running for 2 days I noticed that my one module is no longer receiving messages.

The module logs notes the following:
```
2020-04-07T00:18:32.812603295Z INFO,No IOT message received.
2020-04-07T00:18:39.005434935Z ERROR,ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T00:18:39.005617945Z ERROR,ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation',)
2020-04-07T00:18:39.005778104Z ERROR,UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T00:18:39.005916890Z ERROR,UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation',)
2020-04-07T00:18:39.006027620Z ERROR,Exception caught in background thread.  Unable to handle.
2020-04-07T00:18:39.006213514Z ERROR,["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation',)\n"]
2020-04-07T00:19:02.723699356Z INFO,IOT message received.
2020-04-07T00:22:16.401710573Z INFO,No IOT message received.
2020-04-07T01:16:38.496463054Z ERROR,ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T01:16:38.496880327Z ERROR,ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation',)
2020-04-07T01:16:38.497107579Z ERROR,UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T01:16:38.497272670Z ERROR,UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation',)
2020-04-07T01:16:38.497423074Z ERROR,Exception caught in background thread.  Unable to handle.
2020-04-07T01:16:38.497642965Z ERROR,["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation',)\n"]
2020-04-07T02:14:39.106574400Z ERROR,ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T02:14:39.106931896Z ERROR,ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation',)
2020-04-07T02:14:39.107147802Z ERROR,UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T02:14:39.107260415Z ERROR,UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation',)
2020-04-07T02:14:39.107381175Z ERROR,Exception caught in background thread.  Unable to handle.
2020-04-07T02:14:39.107549129Z ERROR,["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation',)\n"]
2020-04-07T03:12:38.843798834Z ERROR,ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T03:12:38.844217693Z ERROR,ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation',)
2020-04-07T03:12:38.844443777Z ERROR,UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T03:12:38.844584734Z ERROR,UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation',)
2020-04-07T03:12:38.844707624Z ERROR,Exception caught in background thread.  Unable to handle.
2020-04-07T03:12:38.844885474Z ERROR,["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation',)\n"]
2020-04-07T04:10:38.594952545Z ERROR,ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T04:10:38.595298556Z ERROR,ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation',)
2020-04-07T04:10:38.595469670Z ERROR,UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T04:10:38.595631349Z ERROR,UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation',)
2020-04-07T04:10:38.595754787Z ERROR,Exception caught in background thread.  Unable to handle.
2020-04-07T04:10:38.595917935Z ERROR,["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation',)\n"]
2020-04-07T05:08:38.699366654Z ERROR,ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T05:08:38.699583323Z ERROR,ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation',)
2020-04-07T05:08:38.699708498Z ERROR,UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T05:08:38.699828326Z ERROR,UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation',)
2020-04-07T05:08:38.699926888Z ERROR,Exception caught in background thread.  Unable to handle.
2020-04-07T05:08:38.700069587Z ERROR,["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation',)\n"]
2020-04-07T06:06:38.995493191Z ERROR,ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T06:06:38.995705838Z ERROR,ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation',)
2020-04-07T06:06:38.995887761Z ERROR,UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation',)
2020-04-07T06:06:38.996025639Z ERROR,UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation',)
2020-04-07T06:06:38.996153745Z ERROR,Exception caught in background thread.  Unable to handle.
2020-04-07T06:06:38.996327778Z ERROR,["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation',)\n"]
```

The module never recovered from this state, and I had to restart the module. After the restart the module was flooded with messages, thus I can deduce that the edgeHub was still trying to deliver messages to the module while in this state.

In builds prior to v2.1.1 I used to look for the edgeHub debug message "2020-04-07 13:58:37.567 +00:00 [WRN] - Module some-device/some_module is not connected" and if persistent for 15 minutes then I would restart the module from a CRON job on the edge node. This unfortunately can no longer be done, because during this state no messages was output from edgeHub to monitor for this.

I suppose I now will have to come up with a new scheme (for production sites)...

For some more info, I send about 45 messages every second between some of the modules on the edge pipeline. The modules getting the most messages are the most prone to this condition. Modules that only send never has this issue.

@BertKleewein thanks for the watchdog that monitors the disconnection, I can see that it works, but I think there is still another disconnect/issue at play as well.

Thanks for all the efforts.

AB#7366703

**Comment by tanieee28** [comment]
Hi,
did we had any solution for the above problem ?? 
I am facing the same issue with **azure-iot-device-sdk-python**

**Requirement:** I have written a python script using above sdk to send 180 json records in one minute.

**Problem:** the script runs smooth for some hours and send around 180 records, but at some time between it loses the connection to iot hub and hence, the record count sent to iot hub decreases. 

Can i get any help on this ??

Below, are the logs for the time, when the python script lost the connection.

```
Live Generation : INFO : Total active threads : 10 Messaging Threads : 0
Live Generation : INFO : No of Records : 180. Timestamp : ['2020-05-06T20:33:00+00:00']
Live Generation : INFO : Average Message Delivery Time : 0.03502751736158735
Live Generation : ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation')
Live Generation : ConnectionLockStage(ReauthorizeConnectionOperation): failing MQTTPublishOperation op because of error
Live Generation : MQTTPublishOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : ConnectionLockStage(ReauthorizeConnectionOperation): failing MQTTPublishOperation op because of error
Live Generation : MQTTPublishOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : ConnectionLockStage(ReauthorizeConnectionOperation): failing MQTTPublishOperation op because of error
Live Generation : MQTTPublishOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation')
Live Generation : Exception caught in background thread.  Unable to handle.
Live Generation : ["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation')\n"]
Live Generation : INFO : Total active threads : 7 Messaging Threads : 0
Live Generation : ERROR : No of Records : 120. Timestamp : ['2020-05-06T20:34:00+00:00']
Live Generation : INFO : Average Message Delivery Time : 0.5325058757281694

```

**Comment by BertKleewein** [comment]
@LouanDuToitS3  & @tanieee28 , I am working to track this down.  The "transport timeout" error indicates a problem in the Paho (MQTT) error, or below (possibly TLS, TCP, or the network socket itself). 

What can you tell me abourt your operating environments?  What OS, what Python version, what kind of hardware, what kind of network connection, etc.

**Comment by dme-development** [comment]
I am seeing exact the same behavior on a device we are currently working on. We use a IoTHubModuleClient, listening in a thread to receive a module message.

This is a device build with the Jetson Nano Developer Kit.

Operating System: Ubuntu 18.04.3 LTS
Kernel: Linux 4.9.140-tegra
Architecture: arm64

Python 3.7.7

Name: azure-iot-device
Version: 2.1.2

Name: paho-mqtt
Version: 1.5.0

Hope this info helps in tracking down the problem.

**Comment by dschenzer** [comment]
@BertKleewein Did you had a chance to track down this issue?

**Comment by BertKleewein** [comment]
@dschenzer, @tanieee28 , @LouanDuToitS3 , @dme-development  I believe I have a fix for this issue.  The attached wheel has a test fix which resolves some 'transport timeout' issues.  I am moving forward with getting a more robust fix into master.  In the mean time, I'd appreciate hearing if the attached wheel resolves your issues.   

This fix was built from https://github.com/BertKleewein/azure-iot-sdk-python/tree/bertk-release-icm


[azure_iot_device-2.1.2_perf_test-py2.py3-none-any.zip](https://github.com/Azure/azure-iot-sdk-python/files/4685006/azure_iot_device-2.1.2_perf_test-py2.py3-none-any.zip)

**Comment by tanieee28** [comment]
@BertKleewein 
Sorry for late reply, and thanks for working out a possible solution. I will use this and test for some days and will get back to you. Meanwhile, following is the information you requested, if it helps.

Operating System: Ubuntu 18 LTS
Machine : Azure VM
Architecture: arm64

Python 3.7/3.6 (tried on both)

Name: azure-iot-device
Version: 2.1.2/2.2

Name: paho-mqtt
Version: latest

**Comment by BertKleewein** [comment]
@dschenzer, @tanieee28 , @LouanDuToitS3 , @dme-development , we've released [azure-iot-device 2.1.3](https://pypi.org/project/azure-iot-device/2.1.3/) which should resolve this issue.  If you keep having this problem (or similar) in 2.1.3 or later, please open a new issue.

**Comment by az-iot-builder-01** [comment]
@LouanDuToitS3, @tanieee28, @BertKleewein, @dme-development, @dschenzer, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
